sdk/lib/rtl/rangelist.c

```diff
@@ -13,546 +13,1556 @@
 
-    DPRINT("RangeMax: %I64x\n", RangeMax);
-    DPRINT("RangeMin: %I64x\n", RangeMin);
+    DPRINT("RtlFindRange: return STATUS_UNSUCCESSFUL\n");
 
-    *Start = RangeMin;
-
-    return STATUS_SUCCESS;
+    return STATUS_UNSUCCESSFUL;
 }
 
-
-/**********************************************************************
- * NAME							EXPORTED
- * 	RtlFreeRangeList
- *
- * DESCRIPTION
- *	Deletes all ranges in a range list.
+/**
+ * @name RtlFreeRangeList
  *
- * ARGUMENTS
- *	RangeList	Pointer to the range list.
+ * Deletes all ranges in a range list.
  *
- * RETURN VALUE
- *	None
+ * @param RangeList
+ * Pointer to the range list.
  *
- * @implemented
+ * @return
+ * None.
  */
+CODE_SEG("PAGE")
+NTSYSAPI
 VOID
 NTAPI
-RtlFreeRangeList(IN PRTL_RANGE_LIST RangeList)
+RtlFreeRangeList(
+    _In_ PRTL_RANGE_LIST RangeList)
 {
-    PLIST_ENTRY Entry;
-    PRTL_RANGE_ENTRY Current;
-
-    while (!IsListEmpty(&RangeList->ListHead))
-    {
-        Entry = RemoveHeadList(&RangeList->ListHead);
-        Current = CONTAINING_RECORD(Entry, RTL_RANGE_ENTRY, Entry);
+    PRTLP_RANGE_LIST_ENTRY RtlEntry;
+    PRTLP_RANGE_LIST_ENTRY NextRtlEntry;
 
-        DPRINT ("Range start: %I64u\n", Current->Range.Start);
-        DPRINT ("Range end:   %I64u\n", Current->Range.End);
+    PAGED_CODE_RTL();
+    ASSERT(RangeList);
 
-        RtlpFreeMemory(Current, 0);
-    }
+    DPRINT("RtlFreeRangeList: RangeList %p, Count %X\n", RangeList, RangeList->Count);
 
     RangeList->Flags = 0;
     RangeList->Count = 0;
-}
 
+    RtlEntry = CONTAINING_RECORD(RangeList->ListHead.Flink, RTLP_RANGE_LIST_ENTRY, ListEntry);
+    NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
 
-/**********************************************************************
- * NAME							EXPORTED
- * 	RtlGetFirstRange
+    while (&RtlEntry->ListEntry != &RangeList->ListHead)
+    {
+        RemoveEntryList(&RtlEntry->ListEntry);
+        RtlpDeleteRangeListEntry(RtlEntry);
+
+        RtlEntry = NextRtlEntry;
+        NextRtlEntry = RtlpEntryFromLink(RtlEntry->ListEntry.Flink);
+    }
+}
+
+/**
+ * @name RtlGetFirstRange
  *
- * DESCRIPTION
- *	Retrieves the first range of a range list.
+ * Retrieves the first range of a range list.
  *
- * ARGUMENTS
- *	RangeList	Pointer to the range list.
- *	Iterator	Pointer to a user supplied list state buffer.
- *	Range		Pointer to the first range.
+ * @param RangeList
+ * Pointer to the range list.
  *
- * RETURN VALUE
- *	Status
+ * @param Iterator
+ * Pointer to a user supplied list state buffer.
  *
- * @implemented
+ * @param OutRange
+ * Pointer to the first range.
+ *
+ * @return
+ * Returns NTSTATUS.
  */
+CODE_SEG("PAGE")
+NTSYSAPI
 NTSTATUS
 NTAPI
-RtlGetFirstRange(IN PRTL_RANGE_LIST RangeList,
-                 OUT PRTL_RANGE_LIST_ITERATOR Iterator,
-                 OUT PRTL_RANGE *Range)
+RtlGetFirstRange(
+    _In_ PRTL_RANGE_LIST RangeList,
+    _Inout_ PRTL_RANGE_LIST_ITERATOR Iterator,
+    _Out_ PRTL_RANGE * OutRange)
```

Comment:
Keep it in sync with the RTL header:
```suggestion
    _Out_ PRTL_RANGE_LIST_ITERATOR Iterator,
    _Outptr_ PRTL_RANGE *Range)
```
and `OutRange` renamed into `Range`.